From 1d6cfda8e2a7ab5dfac3ea32b3ed6d957677fdda Mon Sep 17 00:00:00 2001
From: Helmut Grohne <helmut@subdivi.de>
Date: Mon, 9 Nov 2015 22:07:52 +0100
Subject: [PATCH 1/2] export a variable DPKG_ROOT

This variable holds the value of instdir. It is supposed to be used in
maintainer scripts. It should be prepended to all paths that are
operated on. Currently, dpkg chroots to the instdir before invoking
maintainer scripts, so when it does that DPKG_ROOT is set to the empty
string. Thus currently, DPKG_ROOT is always empty.
---
 man/dpkg.1   | 10 ++++++++++
 src/main.c   |  2 ++
 src/script.c |  2 ++
 3 files changed, 14 insertions(+)

diff --git a/man/dpkg.1 b/man/dpkg.1
index b47c848..86b5160 100644
--- a/man/dpkg.1
+++ b/man/dpkg.1
@@ -866,6 +866,16 @@ Defined by \fBdpkg\fP on the maintainer script environment to a value
 (\(oq\fB0\fP\(cq or \(oq\fB1\fP\(cq) noting whether debugging has been
 requested (with the \fB\-\-debug\fP option) for the maintainer scripts
 (since dpkg 1.18.4).
+.TP
+.B DPKG_ROOT
+Defined by \fBdpkg\fP on the maintainer script environment to indicate
+which installation to act on. The value is supposed to be prepended to
+any path maintainer scripts operate on. During normal operation, this
+variable is empy. When installing packages into a chroot, dpkg
+normally invokes maintainer scripts using chroot and leaves this
+variable empty. Only when installing into a chroot and the package
+being installed declares support for using this variabe, the chroot
+call is skipped and this variable is non-empty (since dpkg 1.18.5).
 .
 .SH FILES
 .TP
diff --git a/src/main.c b/src/main.c
index 0ff04f1..6440e79 100644
--- a/src/main.c
+++ b/src/main.c
@@ -879,6 +879,8 @@ int main(int argc, const char *const *argv) {
   /* Always set environment, to avoid possible security risks. */
   if (setenv("DPKG_ADMINDIR", admindir, 1) < 0)
     ohshite(_("unable to setenv for subprocesses"));
+  if (setenv("DPKG_ROOT", instdir, 1) < 0)
+    ohshite(_("unable to setenv for subprocesses"));
 
   if (!f_triggers)
     f_triggers = (cipaction->arg_int == act_triggers && *argv) ? -1 : 1;
diff --git a/src/script.c b/src/script.c
index d9514da..6bb121e 100644
--- a/src/script.c
+++ b/src/script.c
@@ -105,6 +105,8 @@ maintscript_pre_exec(struct command *cmd)
 			ohshit(_("admindir must be inside instdir for dpkg to work properly"));
 		if (setenv("DPKG_ADMINDIR", admindir + instdirl, 1) < 0)
 			ohshite(_("unable to setenv for subprocesses"));
+		if (setenv("DPKG_ROOT", "", 1) < 0)
+			ohshite(_("unable to setenv for subprocesses"));
 
 		if (chroot(instdir))
 			ohshite(_("failed to chroot to '%.250s'"), instdir);
-- 
2.8.0.rc3

